rfcs/0145-doc-strings.md

```diff
@@ -0,0 +1,499 @@
+
+[1]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/build-support/trivial-builders/default.nix
+[2]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/stdenv/generic/make-derivation.nix
+[3]: https://github.com/NixOS/nixpkgs/blob/master/nixos/lib/make-disk-image.nix
+
+In general, the format for writing documentation strings is **not formally specified**.
+
+Among the formats encountered in the wild, the one used in `nixpkgs/lib` is the only one intended to be rendered as part of an API documentation, via the nixdoc third-party tool, whose syntax has not been standardized.
+
+### Impossible to differentiate from internal comments
+
+The lack of a formal definition of a doc-comment also means there is no reliable way to distinguish them from internal comments, which makes it impossible to access from doc-comments from tooling. Futhermore it makes it very hard to generate accurate and complete reference documentation.
+
+### References to the problems above
+
+This curated link collection highlights the Nix ecosystem's inconsistencies, a primary focus of this RFC.
+
+#### nixpkgs - comment examples
+
+- [lib/attrsets](https://github.com/NixOS/nixpkgs/blob/5323fbf70331f8a7c47f1b4f49841cf74507f77f/lib/attrsets.nix)
+- [trivial-builders](https://github.com/NixOS/nixpkgs/blob/5323fbf70331f8a7c47f1b4f49841cf74507f77f/pkgs/build-support/trivial-builders/default.nix)
+- [stdenv/mkDerivation](https://github.com/NixOS/nixpkgs/blob/5323fbf70331f8a7c47f1b4f49841cf74507f77f/pkgs/stdenv/generic/make-derivation.nix)
+- [nixos/lib/make-disk-image](https://github.com/NixOS/nixpkgs/blob/5323fbf70331f8a7c47f1b4f49841cf74507f77f/nixos/lib/make-disk-image.nix)
+
+# Design
+[design]: #detailed-design
+
+In the following we give a comprehensive overview to our decision that we've made. Detailed arguments for and against every decision can be found in the [Decisions](#Decisions) section
+
+## CommonMark
+
+**CommonMark according to [RFC 72](https://github.com/NixOS/rfcs/blob/master/rfcs/0072-commonmark-docs.md) is the content of all doc-comments.**
+
+Adopting CommonMark as the content for all doc-comments brings the benefits of widely accepted and understood documentation format in tech projects, while maintaining profitability and consistency within the Nix ecosystem by aligning with existing [NixOS/RFC-72](https://github.com/NixOS/rfcs/blob/master/rfcs/0072-commonmark-docs.md).
+
+## `/** */` is the doc-comment format
+
+The decision to use /** to start a doc-comment ensures a unique distinction from regular comments while still allowing seamless writing without IDE or editor support. A single choice provides the best developer experience and simplifies tooling.
+
+## Placement
+
+**The placement describes the relationship between doc-comments and the documentable node.**
+
+A **documentable node** can be:
+
+- Expression 
+- Binding 
+- Lambda Formal
```

Comment:
I'd like to highlight the concern in https://github.com/NixOS/rfcs/pull/145#discussion_r1374930801 again with more examples:

```nix
{
  f =
    let
      z = someExpensiveComputation;
    in
    /** Docs for x */
    x:
    /** Docs for y */
    y:
    x + 1;
}
```

I can imagine the doc comments for `x` and `y` to be hard to extract properly. They're really part of the function `f`, but not written as part of it.

And very similarly:
```nix
{

  incrementBy =
    /** The amount to increment by */
    amount:
    /** The number to increment */
    x:
    x + amount;
  
  # This is of type Int -> Int, how to document the argument?
  increment = incrementBy 1;
  
  # This works, but requires introducing an extra argument, which sacrifices evaluation performance in current Nix.
  increment =
    /** The number to increment by 1 */
    x:
    incrementBy 1 x;
}
```

How do I attach a doc comment to `increment` without introducing an extra unnecessary argument?

Another example comes [from the `lib.path` library I worked on](https://github.com/NixOS/nixpkgs/blob/3f94ed960a414e31dd38de9a15bec5ed0f21e2fc/lib/path/default.nix#L200-L214):

```nix
  hasPrefix =
    path1:
    assert assertMsg
      (isPath path1)
      "lib.path.hasPrefix: First argument is of type ${typeOf path1}, but a path was expected";
    let
      path1Deconstructed = deconstructPath path1;
    in
      path2:
      assert assertMsg
        (isPath path2)
        "lib.path.hasPrefix: Second argument is of type ${typeOf path2}, but a path was expected";
      let
        path2Deconstructed = deconstructPath path2;
      in
```

The function takes two arguments, but the second argument is only available nested inside an `assert` and a `let in`.

I haven't thought about a better alternative, but I believe these cases might be problematic with the current proposal, since it depends on doc comments always being related to a syntax node, which doesn't necessarily correspond to the exposed interface.

Reply:
Notably we only have this problem because Nix is all _implementation_. There's no interface description language such as libraries, modules and types, which most other languages have. And doc comments should be part of the interface, not the implementation.

Reply:
> Notably we only have this problem because Nix is all implementation.

Yes unfortunately there is no language specification. That is why we decided to use a lot of examples instead. Until the language gets specified.

> I'd like to highlight the concern in https://github.com/NixOS/rfcs/pull/145#discussion_r1374930801 again with more examples

Are those examples inspired by the current nixdoc feature called 'argument documentation'? Argument documentation is removed completely I consider this the best approach. (I'll explain why)

I've corrected the example below.
```nix
{
  f =
    let
      z = someExpensiveComputation;
    in
    /** Docs for the lambda x: y: x+1*/
    x:
    y:
    x + 1;
}
```
The nixdoc feature 'argument documentation' is not specified by this RFC. Indeed one could say that it is removed. There are no arguments in the nix language. Only lambdas, where every lambda has an argument. It makes more sense to use those placements to document the whole lambda. (What you consider argument documentation, is removed, instead we open the possibility for `lambda formals` -> `{a,b,c,...}` to be documentable.

> How do I attach a doc comment to increment without introducing an extra unnecessary argument?

```nix
{
  /** Docs could also be here for convinence (as specified)*/
  incrementBy =
    /**
      A function that increments.
      
      Arguments
      - [amount] The amount to increment by
      - [x] The number that is incremented
    */
    amount:
    x:
    x + amount;
  
  # This is of type Int -> Int, how to document the argument?
  /**
      A function that increments by one
      
      Arguments
      - [x] The number that is incremented
  */
  increment = incrementBy 1;
  
  # This is called eta-wrapping
  # Don't do this. as you already said it has performance impacts. Just document like shown above.
  increment =
    /** The number to increment by 1 */
    x: incrementBy 1 x;
}
```

## Why 'argument documentation' is removed (compared to nixdoc)

Lets say you want to document: 

- an anonymous lambda `a: a` (id function)
- an argument of a lambda ` id = a: a`  (id function)

We now specified that every lambda can have its doc-comment right before the lambda itself

`/**Doc for a: a*/a: a`

If we continue with the `nixdoc` argument documentation this placement could not be used.

`/**Doc for argument a*/a: a`

Both placements are exactly the same. Where we consider the first one to be correct. Please document arguments in the doc-comment. Or alternatively use `{/**Doc for formal a*/a, ...}: `.

Reply:
Maybe we should point out, that this rfc will change the developer experience for writing documentation if you are used to the current nixdoc inspired format. Biggest change might be refactoring the so called 'argument documentation'

Reply:
Also your 'advanced' examples simplifies to:


```nix
/**
Check if a path has another path as prefix

# Arguments
- [path1] document 
- [path2] document
*/
hasPrefix =
    path1:
    assert assertMsg
      (isPath path1)
      "lib.path.hasPrefix: First argument is of type ${typeOf path1}, but a path was expected";
    let
      path1Deconstructed = deconstructPath path1;
    in
      path2:
      assert assertMsg
        (isPath path2)
        "lib.path.hasPrefix: Second argument is of type ${typeOf path2}, but a path was expected";
      let
        path2Deconstructed = deconstructPath path2;
      in
```

Don't expect complex features like automatic infering of arguments, write correct documentation yourself. Currently no tool can handle that for you. Automatically infering arguments and constructing documentation from multiple places is super complex as you already pointed out.

Reply:
That wasn't a great example, you're right. I do still think a piece is missing here, the [previous example](https://github.com/NixOS/rfcs/pull/145#discussion_r1374930801) doesn't seem to work very well. Another example where we might have problems is with functions taking functions, like `lib.fileset.fileFilter`, which already currently has to construct a manual listing for arguments of its predicate argument: https://github.com/NixOS/nixpkgs/blob/85f1ba3e51676fa8cc604a3d863d729026a6b8eb/lib/fileset/default.nix#L284-L326

In such a case documentation writers would have to manually create anchors to make parts linkable, which is probably something we don't want.

While I think this could be doable, I don't think this should be a blocker for this RFC.

Reply:
> That wasn't a great example, you're right. I do still think a piece is missing here, the https://github.com/NixOS/rfcs/pull/145#discussion_r1374930801 doesn't seem to work very well.

Yes you're right. Not everything is intuitively documentable. There are limitations. 
I prefer to thinking of lambdas as enclosed atomic expressions that should document their behavior. It would be nice to somehow also automatically document the returned values. Which is possible, but with certain limitations and it gets very complex if you start involving conditionals, where the documentable-node is only visible under certain conditions.

I'm sure there are solutions to this problem, but they involve some handcrafting of specific tooling on top of this rfc, and some efforts from the user, making those things discoverable.

>  construct a manual listing arguments of its predicate argument: https://github.com/NixOS/nixpkgs/blob/85f1ba3e51676fa8cc604a3d863d729026a6b8eb/lib/fileset/default.nix#L284-L326

As said I'd unify the documentation to be on the lambda only.

Something like:
```
/*
    Filter a file set to only contain files matching some predicate.

     # Arguments
     
     - [predicate]   The predicate ...
     - [fileset]  The file set to filter based on the predicate function
    
  */
  fileFilter =
    predicate:
    fileset:
```

You can utilize the power of markdown, and have an extra section for the `predicate` if it is more complex and requires extra explanation. 

Unfortunately specifying the argument reference syntax is out of scope for this rfc, I'll start another one later to define fundamental references and markdown features. (If it is not a tooling question)

Reply:
I like the bracket syntax for parameters, as it looks like a markdown identifier. A natural "extension" would be to support `[label](identifier)`, effectively renaming the argument for the purpose of documentation.
This is entirely valid because **positional arguments don't actually have a name in the function interface**. Any names we do give them are effectively pseudocode.

Nixdoc does not recognize this fact, causing some developers to apply elaborate renaming to keep the docs clean, even if they need to preprocess an argument (which normally degrades the intended identifier).

I suppose the alternative is to only accept markdown labels and ignore the names used in the implementation. This would seem error-prone - too easy to accidentally switch them up, with no way to check against this in the tooling.

------

Another question: Is it possible to intentionally **document fewer arguments** than the implementation?
For instance, extends is usually thought of as returning a function, rather than being a single curried function.
(The returned function goes from self to self; suitable for passing to fix, which is then responsible for instantiating an overlay. Docs are WIP)

```
  extends = f: rattrs: self: let super = rattrs self; in super // f self super;
```

But typical use is along the lines of

```
  fix (extends f g)
```

Reply:
> I like the bracket syntax for parameters, as it looks like a markdown identifier. 

:heart: 

> A natural "extension" would be to support [label](identifier)
> ...
> Is it possible to intentionally document fewer arguments than the implementation?

Those questions are not answered by the RFC. It doesn't include any markdown extensions or styling rules. It is unfortunate. I try to keep the scope manageable and focused on the minimal requirements, although I always have to resist the temptation not to define those nice (to have) features.

For now i think this is a tooling question, whether the blessed tools supports certain extensions.


> positional arguments don't actually have a name in the function interface

Yes that bothers me too. Often our documentation refers to semantic names, which is wrong. It should correctly be the order in which the arguments must be supplied, from the perspective of the user.  This means if `self` is not passed or accessed by the caller. It should not be documented on the interface.

Reply:
@infinisil. Are your concerns resolved? 

I put a sentence in the migration section, that this RFC uses a different "mental mode"l than the former `nixdoc` one.

> Note: The current nixdoc feature 'Function arguments' uses a different mental model than this RFC. Arguments must now be explicitly documented inside of the lambda documentation.